Handle slices in  `mlx_funcify_IncSubtensor`

jessegrabowski opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-devs/pytensor/releases -->

## Description
The MLX dispatch for `IncSubtensor` was assuming that the indexes would always be integers, but they can actually be either integers or slices. This PR adds logic to handle the slice case.

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [ ] Closes #1690 
- [ ] Related to #1350 

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [ ] New feature / enhancement
- [x] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->


<!-- readthedocs-preview pytensor start -->
----
📚 Documentation preview 📚: https://pytensor--1692.org.readthedocs.build/en/1692/

<!-- readthedocs-preview pytensor end -->

jessegrabowski committed:
Handle slices in  `mlx_funcify_IncSubtensor`

jessegrabowski commented on pytensor/link/mlx/dispatch/subtensor.py:
```diff
@@ -64,7 +64,25 @@ def mlx_fn(x, indices, y):
             return x
 
     def incsubtensor(x, y, *ilist, mlx_fn=mlx_fn, idx_list=idx_list):
-        indices = indices_from_subtensor(ilist, idx_list)
```
the traceback is in the linked issue, does it answer your question?

ricardoV94 replied:
The traceback makes me suspect that's it but doesn't answer because it doesn't say what's the actual slice element (or type) that it's complaining about

ricardoV94 commented:
So we should open an issue with them and I'm surprised any indexing worked before. We're casting constant numpy integers to python int? If so we should do something for the tipefy of slices

cetagostini committed:
Create test_mlx_indexing_behavior.py

cetagostini committed:
Fix MLX indexing to support np.int64 and NumPy scalars

Introduces normalize_indices_for_mlx to convert NumPy integer and floating types, MLX scalar arrays, and slice components to Python int/float for MLX compatibility. Updates all MLX subtensor dispatch functions to use this normalization, resolving issues with MLX's strict indexing requirements. Adds comprehensive tests for np.int64 indices and slices in subtensor and inc_subtensor operations, including advanced indexing scenarios.

cetagostini committed:
Add missing newline at end of files

Appended a newline to the end of subtensor.py and test_subtensor.py to conform with POSIX standards and improve code consistency.

ricardoV94 commented:
Yeah we know it. I'll take a look to stop it but it need not block us

ricardoV94 commented on tests/link/mlx/test_mlx_indexing_behavior.py:
```diff
@@ -0,0 +1,212 @@
+mx = pytest.importorskip("mlx.core")
+
+
+def test_mlx_python_int_indexing():
```
why would we be testing mlx directly?

cetagostini replied:
Ah not at all this was internal for me, let me delete this file. I didn't wanna to push!

cetagostini commented on pytensor/link/mlx/dispatch/subtensor.py:
```diff
@@ -13,12 +15,119 @@
 from pytensor.tensor.type_other import MakeSlice
 
 
+def normalize_indices_for_mlx(ilist, idx_list):
```
So, we should simply close this pr?

cetagostini replied:
You're probably right that it's overkill. I wanted to make sure I was handling all the edge cases, but I might have over-engineered it 😅

Just to make sure I understand - are you suggesting something like this would be enough?

```python
def subtensor(x, ilists):
 # Convert mx.array scalars and np.int64 to Python int before indices_from_subtensor
 normalized = tuple(
 int(e.item()) if isinstance(e, mx.array) and e.ndim == 0
 else int(e) if isinstance(e, np.integer)
 else e
 for e in ilists
 )
 indices = indices_from_subtensor(normalized, idx_list)
 if len(indices) == 1:
 indices = indices[0]
 return x.getitem(indices)
```

The only thing that made me hesitant about this approach: I wasn't 100% sure if ilists always contains just scalar values, or if we might sometimes get already-constructed slice objects with mx.array components. If it's always scalars going into indices_from_subtensor, then yeah, normalizing beforehand is way cleaner!

Happy to simplify if you confirm this approach would cover all the cases. The current version works, but simpler is always better if it's safe.

ricardoV94 replied:
No, it should just be way less code from what I could gather.

The helper is fine if the AdvancedSubtensor also needs it, but the docstrings need only say we need to convert numpy integers to python integers (which btw will fail with symbolic indices but that's a different matter).

cetagostini committed:
modify

cetagostini committed:
Simplify based on Ricardo input

cetagostini committed:
Update subtensor.py

cetagostini committed:
Simplify error msg

ricardoV94 commented on pytensor/link/mlx/dispatch/subtensor.py:
```diff
@@ -13,12 +16,51 @@
+        else:
+            return element
+
+    indices = indices_from_subtensor(ilist, idx_list)
```
move this outside the helper, as it is only relevant for basic Subtensor/IncSubtensor. The advanced methods don't have an `idx_list` and don't need `indices_from_subtensor`

ricardoV94 commented on pytensor/link/mlx/dispatch/subtensor.py:
```diff
@@ -30,10 +72,12 @@ def subtensor(x, *ilists):
 @mlx_funcify.register(AdvancedSubtensor1)
 def mlx_funcify_AdvancedSubtensor(op, node, **kwargs):
+    """MLX implementation of AdvancedSubtensor."""
     idx_list = getattr(op, "idx_list", None)
```
Not a thing for Advanced indexing
```suggestion
```

cetagostini commented on pytensor/link/mlx/dispatch/subtensor.py:
```diff
@@ -13,12 +16,51 @@
 @mlx_funcify.register(Subtensor)
 def mlx_funcify_Subtensor(op, node, **kwargs):
+    """MLX implementation of Subtensor."""
     idx_list = getattr(op, "idx_list", None)
```
Done!

cetagostini committed:
Refactor MLX subtensor dispatch and update test

Simplifies index normalization logic in MLX subtensor dispatch functions by separating basic and advanced indexing cases. Updates the advanced incsubtensor test to use vector array indices and a matching value shape for improved coverage.

cetagostini committed:
pushing a few more tests.

jessegrabowski commented:
@ricardoV94 @cetagostini was there anything else left to do here?

ricardoV94 commented:
Is this aligned with the new subtensor API (in v3)? Otherwise it's going to be double work
